File: 0174-dungeon-game/0174-dungeon-game.java
class Solution {
  public int calculateMinimumHP(int[][] dungeon) {
    int m = dungeon.length;
    int n = dungeon[0].length;
    int[][] dp = new int[m][n];
    dp[m - 1][n - 1] = Math.max(1, 1 - dungeon[m - 1][n - 1]);
    
    for (int i = n - 2; i >= 0; i--) {
      dp[m - 1][i] = Math.max(1, dp[m - 1][i + 1] - dungeon[m - 1][i]);
    }
    
    for (int i = m - 2; i >= 0; i--) {
      dp[i][n - 1] = Math.max(1, dp[i + 1][n - 1] - dungeon[i][n - 1]);
    }
    
    for (int r = m - 2; r >= 0; r--) {
      for (int c = n - 2; c >= 0; c--) {
        dp[r][c] = Math.max(1, Math.min(Math.max(1, dp[r][c + 1] - dungeon[r][c]), Math.max(1, dp[r + 1][c] - dungeon[r][c])));
      }
    }
    return dp[0][0];
  }
}

File: 0174-dungeon-game/0174-dungeon-game.py
class Solution:
  def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
    m, n = len(dungeon), len(dungeon[0])
    dp = [[0]*n for _ in range(m)]
    dp[-1][-1] = max(1, 1-dungeon[-1][-1])
    for i in range(n-2, -1, -1):
      dp[-1][i] = max(1, dp[-1][i+1]-dungeon[-1][i])
    for i in range(m-2, -1, -1):
      dp[i][-1] = max(1, dp[i+1][-1]-dungeon[i][-1])
    for r in range(m-2, -1, -1):
      for c in range(n-2, -1, -1):
        dp[r][c] = max(1, min(max(1, dp[r][c+1]-dungeon[r][c]), max(1, dp[r+1][c]-dungeon[r][c])))
    return dp[0][0]